fdb-extensions/src/main/java/com/apple/foundationdb/linear/FhtKacRotator.java

```diff
@@ -0,0 +1,238 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.linear;
+
+import javax.annotation.Nonnull;
+import java.util.Arrays;
+import java.util.Random;
+
+/** FhtKac-like random orthogonal rotator.
+ *  - R rounds (default 4)
+ *  - Per round: random ±1 -> FWHT on largest 2^k block (head/tail alternation) -> π/4 Givens across halves
+ *  Time per apply: O(R * (n log n)) with tiny constants; memory: O(R * n) bits for signs.
+ */
+@SuppressWarnings({"checkstyle:MethodName", "checkstyle:MemberName"})
+public final class FhtKacRotator implements LinearOperator {
+    private final int numDimensions;
+    private final int rounds;
+    private final byte[][] signs; // signs[r][i] in {-1, +1}
+    private static final double INV_SQRT2 = 1.0 / Math.sqrt(2.0);
+
+    public FhtKacRotator(final long seed, final int numDimensions, final int rounds) {
+        if (numDimensions < 2) {
+            throw new IllegalArgumentException("n must be >= 2");
+        }
+        if (rounds < 1) {
+            throw new IllegalArgumentException("rounds must be >= 1");
+        }
+        this.numDimensions = numDimensions;
+        this.rounds = rounds;
+
+        // Pre-generate Rademacher signs for determinism/reuse.
+        final Random rng = new Random(seed);
+        this.signs = new byte[rounds][numDimensions];
+        for (int r = 0; r < rounds; r++) {
+            for (int i = 0; i < numDimensions; i++) {
+                signs[r][i] = rng.nextBoolean() ? (byte)1 : (byte)-1;
+            }
+        }
+    }
+
+    @Override
+    public int getRowDimension() {
+        return numDimensions;
+    }
+
+    @Override
+    public int getColumnDimension() {
+        return numDimensions;
+    }
+
+    @Override
+    public boolean isTransposable() {
+        return true;
+    }
+
+    @Nonnull
+    @Override
+    public RealVector operate(@Nonnull final RealVector x) {
+        return new DoubleRealVector(operate(x.getData()));
+    }
+
+    @Nonnull
+    private double[] operate(@Nonnull final double[] x) {
+        if (x.length != numDimensions) {
+            throw new IllegalArgumentException("dimensionality of x != n");
+        }
+        final double[] y = Arrays.copyOf(x, numDimensions);
+
+        for (int r = 0; r < rounds; r++) {
+            // 1) Rademacher signs
+            byte[] s = signs[r];
+            for (int i = 0; i < numDimensions; i++) {
+                y[i] = (s[i] == 1 ? y[i] : -y[i]);
+            }
+
+            // 2) FWHT on largest 2^k block; alternate head/tail
```

Comment:
What is FWHT?

Reply:
Fast Walsh–Hadamard Transform. I removed all references to Fast Hadamard Transform (FHT) to match the class name.